pkg/retriever/bitswaphelpers/preloadcachingstorage.go

```diff
@@ -0,0 +1,368 @@
+		// load from cache, write direct to parent and return
+		r, err := linkSystemGetStream(cs.cacheLinkSystem, linkCtx, link)
+		if err != nil {
+			return nil, err
+		}
+		return loadTo(
+			r,
+			[]linking.BlockWriteOpener{cs.parentLinkSystem.StorageWriteOpener},
+			linkCtx,
+			link)
+	}
+}
+
+func loadTo(
+	reader io.Reader,
+	destination []linking.BlockWriteOpener,
+	linkCtx linking.LinkContext,
+	link ipld.Link) (io.Reader, error) {
+
+	blc := make([]linking.BlockWriteCommitter, 0, len(destination))
+	for _, blo := range destination {
+		writer, c, err := blo(linkCtx)
+		if err != nil {
+			return nil, err
+		}
+		blc = append(blc, c)
+		reader = io.TeeReader(reader, writer)
+	}
+	byts, err := io.ReadAll(reader) // slurp it in so the writers can commit
+	if err != nil {
+		return nil, err
+	}
+	err = nil
+	for _, c := range blc {
+		err = multierr.Append(err, c(link))
+	}
+	return bytes.NewBuffer(byts), err
+}
+
+func (cs *PreloadCachingStorage) preloadLink(pl *preloadingLink, linkCtx linking.LinkContext, link ipld.Link) {
+	pl.loadSyncer.Do(func() {
+		defer close(pl.loaded)
+		reader, err := cs.fetcher(linkCtx, link)
+		if err != nil {
+			if nf, ok := err.(interface{ NotFound() bool }); ok && nf.NotFound() {
+				cs.preloadsLk.Lock()
+				cs.notFound[string(link.(cidlink.Link).Cid.Hash())] = struct{}{}
+				cs.preloadsLk.Unlock()
+			}
+			pl.err = err
+		} else {
+			w, c, err := cs.cacheLinkSystem.StorageWriteOpener(linkCtx)
+			if err != nil {
+				pl.err = err
+				return
+			}
+			if _, err := io.Copy(w, reader); err != nil {
+				pl.err = err
+				return
+			}
+			if err := c(link); err != nil {
+				pl.err = err
+				return
+			}
+		}
+	})
+}
+
+func (cs *PreloadCachingStorage) Start(ctx context.Context) {
+	if cs.cancel != nil {
+		panic("already started")
+	}
+	ctx, cs.cancel = context.WithCancel(ctx)
+	go cs.run(ctx)
+}
+
+func (cs *PreloadCachingStorage) run(ctx context.Context) {
+	feed := make(chan *request)
+	defer func() {
+		//close(feed)
+	}()
+
+	for i := 0; i < cs.concurrency; i++ {
+		go func() {
+			for request := range feed {
+				cs.preloadsLk.RLock()
+				pl, ok := cs.preloads[request.link]
+				cs.preloadsLk.RUnlock()
+				if !ok {
+					continue
+				}
+				cs.preloadLink(pl, request.linkCtx, request.link)
+			}
+		}()
+	}
+
+	requestBuffer := list.New()
+	var send chan<- *request
+	var next *request
+	for {
+		select {
+		case request := <-cs.requests:
+			if next == nil {
+				next = &request
+				send = feed
+			} else {
+				requestBuffer.PushBack(&request)
+			}
+		case send <- next:
+			if requestBuffer.Len() > 0 {
+				next = requestBuffer.Remove(requestBuffer.Front()).(*request)
+			} else {
+				next = nil
+				send = nil
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (cs *PreloadCachingStorage) Stop() error {
+	if cs.cancel != nil {
+		cs.cancel()
+		cs.cancel = nil
+	}
+	return nil
+}
+
+// TODO: remove this and the stats it uses
```

Comment:
Is this worth just making a metric?

Reply:
Cleaned this up a bit and added a `BitswapPreloadedHitFraction` that's a distribution metric between 0 and 1 that should tells us how many blocks were fully fetched by the time we came back and needed them. I've added a bunch of docs with some of my thoughts about how the stats could be read.

In `TestHttpFetch/bitswap_large_sharded_directory` we get ~0.7, which is very nice.